[SIMD] Intel support for load and store operations and fix operation shuffle

hyjorc1 opened the pull request:
#### 0def43da76cfb0f0fce99e5ff7a3e2b57f8e7dce
<pre>
[SIMD] Intel support for load and store operations and fix operation shuffle
<a href="https://bugs.webkit.org/show_bug.cgi?id=249073">https://bugs.webkit.org/show_bug.cgi?id=249073</a>
rdar://103214919

Reviewed by Yusuke Suzuki.

This patch achieves four tasks:
1. Add load and store operations for WASM SIMD on Intel.
   <a href="https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#load-and-store">https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#load-and-store</a>
2. And fix WASM SIMD operation i8x16.shuffle.
3. Fix instruction encoding in X86Assember.h according to the new conventions.
4. Define and use AVX functions for operations replace_lane, extract_lane, and splat.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorReplaceLaneAVX):
(JSC::MacroAssemblerX86_64::vectorReplaceLane):
(JSC::MacroAssemblerX86_64::vectorExtractLane):
(JSC::MacroAssemblerX86_64::vectorExtractLaneAVX):
(JSC::MacroAssemblerX86_64::vectorAbs):
(JSC::MacroAssemblerX86_64::vectorTruncSatUnsignedFloat64):
(JSC::MacroAssemblerX86_64::vectorExtendHigh):
(JSC::MacroAssemblerX86_64::vectorSplatAVX):
(JSC::MacroAssemblerX86_64::vectorSplat):
(JSC::MacroAssemblerX86_64::vectorLoad8Splat):
(JSC::MacroAssemblerX86_64::vectorLoad16Splat):
(JSC::MacroAssemblerX86_64::vectorLoad32Splat):
(JSC::MacroAssemblerX86_64::vectorLoad64Splat):
(JSC::MacroAssemblerX86_64::vectorLoad8Lane):
(JSC::MacroAssemblerX86_64::vectorLoad16Lane):
(JSC::MacroAssemblerX86_64::vectorLoad32Lane):
(JSC::MacroAssemblerX86_64::vectorLoad64Lane):
(JSC::MacroAssemblerX86_64::vectorStore8Lane):
(JSC::MacroAssemblerX86_64::vectorStore16Lane):
(JSC::MacroAssemblerX86_64::vectorStore32Lane):
(JSC::MacroAssemblerX86_64::vectorStore64Lane):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pinsrb_i8rr):
(JSC::X86Assembler::pinsrw_i8rr):
(JSC::X86Assembler::pinsrd_i8rr):
(JSC::X86Assembler::pinsrq_i8rr):
(JSC::X86Assembler::insertps_i8rr):
(JSC::X86Assembler::pextrb_i8rr):
(JSC::X86Assembler::pextrw_i8rr):
(JSC::X86Assembler::pextrd_i8rr):
(JSC::X86Assembler::pextrq_i8rr):
(JSC::X86Assembler::pshufd_i8rr):
(JSC::X86Assembler::pshuflw_i8rr):
(JSC::X86Assembler::shufps_i8rr):
(JSC::X86Assembler::shufpd_i8rr):
(JSC::X86Assembler::pblendw_rr):
(JSC::X86Assembler::packusdw_rr):
(JSC::X86Assembler::pmovsxbw):
(JSC::X86Assembler::pmovzxbw):
(JSC::X86Assembler::pmovsxwd):
(JSC::X86Assembler::pmovzxwd):
(JSC::X86Assembler::pmovsxdq):
(JSC::X86Assembler::pmovzxdq):
(JSC::X86Assembler::vpinsrb_i8mrr):
(JSC::X86Assembler::vpinsrb_i8rrr):
(JSC::X86Assembler::vpinsrw_i8mrr):
(JSC::X86Assembler::vpinsrw_i8rrr):
(JSC::X86Assembler::vpinsrd_i8mrr):
(JSC::X86Assembler::vpinsrd_i8rrr):
(JSC::X86Assembler::vpinsrq_i8mrr):
(JSC::X86Assembler::vpinsrq_i8rrr):
(JSC::X86Assembler::vinsertps_i8rrr):
(JSC::X86Assembler::vmovddup_rr):
(JSC::X86Assembler::vmovddup_mr):
(JSC::X86Assembler::vmovapd_rr):
(JSC::X86Assembler::vbroadcastss_mr):
(JSC::X86Assembler::vmovq_rr):
(JSC::X86Assembler::vpunpcklqdq_rrr):
(JSC::X86Assembler::vunpcklpd_rrr):
(JSC::X86Assembler::vpextrb_i8rr):
(JSC::X86Assembler::vpextrb_i8rm):
(JSC::X86Assembler::vpextrw_i8rr):
(JSC::X86Assembler::vpextrw_i8rm):
(JSC::X86Assembler::vpextrd_i8rr):
(JSC::X86Assembler::vpextrd_i8rm):
(JSC::X86Assembler::vpextrq_i8rr):
(JSC::X86Assembler::vpextrq_i8rm):
(JSC::X86Assembler::vshufps_i8rrr):
(JSC::X86Assembler::vshufpd_i8rrr):
(JSC::X86Assembler::vpshuflw_i8rr):
(JSC::X86Assembler::vpshufd_i8rr):
(JSC::X86Assembler::vpackusdw_rrr):
(JSC::X86Assembler::vpmovsxbw_rr):
(JSC::X86Assembler::vpmovzxbw_rr):
(JSC::X86Assembler::vpmovsxwd_rr):
(JSC::X86Assembler::vpmovzxwd_rr):
(JSC::X86Assembler::vpmovsxdq_rr):
(JSC::X86Assembler::vpmovzxdq_rr):
(JSC::X86Assembler::vmovaps_rr):
(JSC::X86Assembler::vmovshdup_rr):
(JSC::X86Assembler::vmovhlps_rrr):
(JSC::X86Assembler::vmovsd_rrr):
(JSC::X86Assembler::vpblendw_i8rrr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
(JSC::X86Assembler::pinsrb_rr): Deleted.
(JSC::X86Assembler::pinsrw_rr): Deleted.
(JSC::X86Assembler::pinsrd_rr): Deleted.
(JSC::X86Assembler::pinsrq_rr): Deleted.
(JSC::X86Assembler::insertps_rr): Deleted.
(JSC::X86Assembler::pextrb_rr): Deleted.
(JSC::X86Assembler::pextrw_rr): Deleted.
(JSC::X86Assembler::pextrd_rr): Deleted.
(JSC::X86Assembler::pextrq_rr): Deleted.
(JSC::X86Assembler::pshufd_rr): Deleted.
(JSC::X86Assembler::pshuflw_rr): Deleted.
(JSC::X86Assembler::shufps_rr): Deleted.
(JSC::X86Assembler::shufpd_rr): Deleted.
(JSC::X86Assembler::vpextrb_rr): Deleted.
(JSC::X86Assembler::vpextrw_rr): Deleted.
(JSC::X86Assembler::vpextrd_rr): Deleted.
(JSC::X86Assembler::vpextrq_rr): Deleted.
(JSC::X86Assembler::vshufps_rrr): Deleted.
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSwizzleHelperX86):
(JSC::Wasm::AirIRGenerator::addSIMDV_VV):
(JSC::Wasm::AirIRGenerator::addSIMDShuffle):
(JSC::Wasm::AirIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::AirIRGenerator::addSIMDSwizzle): Deleted.

Canonical link: <a href="https://commits.webkit.org/257818@main">https://commits.webkit.org/257818@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c5a1522fc26fb879a3817325e52b03509d31222

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100123 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9292 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33197 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109458 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/169693 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10175 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92555 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107348 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105892 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7689 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90706 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3062 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86668 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/491 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3032 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29048 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9162 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43365 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89549 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5370 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/4872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20021 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3299,15 +3299,77 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+    {
+        load64(address, scratch);
+        vectorSplat64(scratch, dest);
+    }
```
I think we can implement them more efficiently. For example, how about using vmmoddup_mr?

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3299,15 +3299,77 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+    void vectorLoad8Lane(Address address, TrustedImm32 imm, FPRegisterID dest, RegisterID scratch)
+    {
+        load8(address, scratch);
+        vectorReplaceLane(SIMDLane::i8x16, imm, scratch, dest);
```
Ditto. I think they can be implemented in a better way with memory operand.

Constellation reviewed (changes requested):
Commented

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2501,6 +2502,16 @@ class X86Assembler {
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, OP3_PINSRB, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        m_formatter.immediate8(laneIndex);
+    }
+
```
Place them in the right part in X86Assembler. This is not the place used for defining AVX functions.
Please strictly follow to the style in X86Assembler to keep consistency.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2511,6 +2522,16 @@ class X86Assembler {
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_66, OP2_PINSRW_VdqRdqp, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        m_formatter.immediate8(laneIndex);
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2521,6 +2542,16 @@ class X86Assembler {
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, OP3_PINSRD, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        m_formatter.immediate8(laneIndex);
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2531,6 +2562,16 @@ class X86Assembler {
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, VexW::W1, OP3_PINSRD, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
+        m_formatter.immediate8(laneIndex);
+    }
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -3762,6 +3803,22 @@ class X86Assembler {
+        // A	NA	ModRM:reg (w)	ModRM:r/m (r)	NA	NA
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp38, OP3_VBROADCASTSS, (RegisterID)dst, (RegisterID)0, base, offset);
+    }
+
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -6297,6 +6359,15 @@ class X86Assembler {
             writer.registerModRM(dest, b);
         }
 
+        void vexNdsLigThreeByteOp(OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, VexW vexW, ThreeByteOpcodeID opcode, RegisterID dest, RegisterID a, RegisterID b, int offset)
```
Use `base` instead of `b`.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -6286,6 +6343,11 @@ class X86Assembler {
             vexNdsLigThreeByteOp(simdPrefix, impliedBytes, VexW::W0, opcode, dest, a, b);
         }
 
+        void vexNdsLigWigThreeByteOp(OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, ThreeByteOpcodeID opcode, RegisterID dest, RegisterID a, RegisterID b, int offset)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        if (supportsAVXForSIMD())
+            m_assembler.vpinsrq_mrr(imm.m_value, address.offset, address.base, dest, dest);
+        else
+            RELEASE_ASSERT_NOT_REACHED();
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad32Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad16Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad8Lane(Address address, TrustedImm32 imm, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad64Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad32Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD())
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+
+    void vectorLoad16Splat(Address address, FPRegisterID dest)
+    {
+        if (supportsAVXForSIMD()) {
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
-    void vectorLoad64Lane(Address address, TrustedImm32 imm, FPRegisterID dest) { UNUSED_PARAM(address); UNUSED_PARAM(imm); UNUSED_PARAM(dest); }
+    void vectorLoad8Splat(Address address, FPRegisterID dest, FPRegisterID scratch)
+    {
+        if (supportsAVXForSIMD()) {
```
Let's just make it `ASSERT(supportsAVXForSIMD())`

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+            m_assembler.vpinsrw_mrr(0, address.offset, address.base, dest, dest);
+            m_assembler.pshuflw_rr(0, dest, dest);
+            m_assembler.punpcklqdq_rr(dest, dest);
+        } else
```
Define and use AVX functions.

hyjorc1 replied:
For other AVX functions, I will implement them in another patch.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3297,97 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+    void vectorStore8Lane(TrustedImm32 imm, FPRegisterID src, Address address, RegisterID scratch)
+    {
+        vectorExtractLane(SIMDLane::i8x16, SIMDSignMode::Unsigned, imm, src, scratch);
+        store8(scratch, address);
```
Do we need these two separate things? I think vpextrb etc. can take memory destination, so we can remove scratch etc. too. and efficiently implement them.

Constellation replied:
Like, defining `vpextrb_rm` and use it.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_66, OP2_MOVAPD_VpdWpd, (RegisterID)dst, (RegisterID)0, (RegisterID)src);
+    }
+
+    void vbroadcast_mr(int offset, RegisterID base, XMMRegisterID dst)
```
There is no vbroadcast_mr instruction. `VBROADCASTSS`

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/movd:movq
+        // VEX.128.66.0F.W1 6E /r VMOVQ xmm1, r64/m64
+        // A	NA	ModRM:reg (w)	ModRM:r/m (r)	NA	NA
+        m_formatter.vexNdsLigWigTwoByteOp(PRE_SSE_66, VexW::W1, OP2_MOVD_VdEd, (RegisterID)dest, (RegisterID)0, (RegisterID)src);
```
Define MOVQ opcode instead of using OP2_MOVD_VdEd.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/vbroadcast
+        // VEX.128.66.0F38.W0 18 /r VBROADCASTSS xmm1, m32
+        // A	NA	ModRM:reg (w)	ModRM:r/m (r)	NA	NA
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp38, OP3_VBROADCASTSS, (RegisterID)dst, (RegisterID)0, base, offset);
```
Don't `OP3_VBROADCASTSS` have no arguments?

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/pinsrb:pinsrd:pinsrq
+        // VEX.128.66.0F3A.W1 22 /r ib VPINSRQ xmm1, xmm2, r/m64, imm8
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, VexW::W1, OP3_PINSRD, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
```
OP3_PINSRD is not right name for `VPINSRQ` op.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/pinsrb:pinsrd:pinsrq
+        // VEX.128.66.0F3A.W0 22 /r ib VPINSRD xmm1, xmm2, r/m32, imm8
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, OP3_PINSRD, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
```
Doesn't `OP3_PINSRD` take arguments?

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4330,6 +4331,94 @@ class X86Assembler {
+        // https://www.felixcloutier.com/x86/pinsrb:pinsrd:pinsrq
+        // VEX.128.66.0F3A.W0 20 /r ib VPINSRB xmm1, xmm2, r32/m8, imm8
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)	imm8
+        m_formatter.vexNdsLigWigThreeByteOp(PRE_SSE_66, VexImpliedBytes::ThreeBytesOp3A, OP3_PINSRB, (RegisterID)xmm1, (RegisterID)xmm2, (RegisterID)base, offset);
```
Doesn't `OP3_PINSRB` take arguments?

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3338,81 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+        ASSERT(supportsAVXForSIMD());
+        m_assembler.vpinsrw_i8mrr(0, address.offset, address.base, dest, dest);
+        m_assembler.pshuflw_i8rr(0, dest, dest);
+        m_assembler.punpcklqdq_rr(dest, dest);
```
They are not AVX functions. FIXME is necessary.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -3297,15 +3338,81 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
+    void vectorLoad32Splat(Address address, FPRegisterID dest)
+    {
+        ASSERT(supportsAVXForSIMD());
+        m_assembler.vbroadcast_mr(address.offset, address.base, dest);
```
`vbroadcast` does not exist. It is `vbroadcastss_rr`

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5134,6 +5351,38 @@ class X86Assembler {
+    {
+        // https://www.felixcloutier.com/x86/movaps
+        // VEX.128.0F.WIG 28 /r VMOVAPS xmm1, xmm2/m128
+        // A	NA	ModRM:reg (w)	ModRM:r/m (r)	NA	NA
```
NANA => NA NA
And let's use 4 spaces for tabs in this line to align it to the other implementation. Let's keep consistency.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5134,6 +5351,38 @@ class X86Assembler {
+    {
+        // https://www.felixcloutier.com/x86/movshdup
+        // VEX.128.F3.0F.WIG 16 /r VMOVSHDUP xmm1, xmm2/m128
+        // A	NA	ModRM:reg (w)	ModRM:r/m (r)	NA	NA
```
Ditto

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5134,6 +5351,38 @@ class X86Assembler {
+    {
+        // https://www.felixcloutier.com/x86/movhlps
+        // VEX.128.0F.WIG 12 /r VMOVHLPS xmm1, xmm2, xmm3
+        // RVM	ModRM:reg (w)	vvvv (r)	ModRM:r/m (r)	NA
```
Ditto

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -5134,6 +5351,38 @@ class X86Assembler {
+    {
+        // https://www.felixcloutier.com/x86/movsd
+        // VEX.LIG.F2.0F.WIG 10 /r VMOVSD xmm1, xmm2, xmm3
+        // B	NA	ModRM:reg (w)	VEX.vvvv (r)	ModRM:r/m (r)
```
Ditto

Constellation reviewed (approved):
r=me with comments

hyjorc1 committed:
[SIMD] Intel support for load and store operations and fix operation shuffle
https://bugs.webkit.org/show_bug.cgi?id=249073
rdar://103214919

Reviewed by Yusuke Suzuki.

This patch achieves four tasks:
1. Add load and store operations for WASM SIMD on Intel.
   https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#load-and-store
2. And fix WASM SIMD operation i8x16.shuffle.
3. Fix instruction encoding in X86Assember.h according to the new conventions.
4. Define and use AVX functions for operations replace_lane, extract_lane, and splat.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorReplaceLaneAVX):
(JSC::MacroAssemblerX86_64::vectorReplaceLane):
(JSC::MacroAssemblerX86_64::vectorExtractLane):
(JSC::MacroAssemblerX86_64::vectorExtractLaneAVX):
(JSC::MacroAssemblerX86_64::vectorAbs):
(JSC::MacroAssemblerX86_64::vectorTruncSatUnsignedFloat64):
(JSC::MacroAssemblerX86_64::vectorExtendHigh):
(JSC::MacroAssemblerX86_64::vectorSplatAVX):
(JSC::MacroAssemblerX86_64::vectorSplat):
(JSC::MacroAssemblerX86_64::vectorLoad8Splat):
(JSC::MacroAssemblerX86_64::vectorLoad16Splat):
(JSC::MacroAssemblerX86_64::vectorLoad32Splat):
(JSC::MacroAssemblerX86_64::vectorLoad64Splat):
(JSC::MacroAssemblerX86_64::vectorLoad8Lane):
(JSC::MacroAssemblerX86_64::vectorLoad16Lane):
(JSC::MacroAssemblerX86_64::vectorLoad32Lane):
(JSC::MacroAssemblerX86_64::vectorLoad64Lane):
(JSC::MacroAssemblerX86_64::vectorStore8Lane):
(JSC::MacroAssemblerX86_64::vectorStore16Lane):
(JSC::MacroAssemblerX86_64::vectorStore32Lane):
(JSC::MacroAssemblerX86_64::vectorStore64Lane):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pinsrb_i8rr):
(JSC::X86Assembler::pinsrw_i8rr):
(JSC::X86Assembler::pinsrd_i8rr):
(JSC::X86Assembler::pinsrq_i8rr):
(JSC::X86Assembler::insertps_i8rr):
(JSC::X86Assembler::pextrb_i8rr):
(JSC::X86Assembler::pextrw_i8rr):
(JSC::X86Assembler::pextrd_i8rr):
(JSC::X86Assembler::pextrq_i8rr):
(JSC::X86Assembler::pshufd_i8rr):
(JSC::X86Assembler::pshuflw_i8rr):
(JSC::X86Assembler::shufps_i8rr):
(JSC::X86Assembler::shufpd_i8rr):
(JSC::X86Assembler::pblendw_rr):
(JSC::X86Assembler::packusdw_rr):
(JSC::X86Assembler::pmovsxbw):
(JSC::X86Assembler::pmovzxbw):
(JSC::X86Assembler::pmovsxwd):
(JSC::X86Assembler::pmovzxwd):
(JSC::X86Assembler::pmovsxdq):
(JSC::X86Assembler::pmovzxdq):
(JSC::X86Assembler::vpinsrb_i8mrr):
(JSC::X86Assembler::vpinsrb_i8rrr):
(JSC::X86Assembler::vpinsrw_i8mrr):
(JSC::X86Assembler::vpinsrw_i8rrr):
(JSC::X86Assembler::vpinsrd_i8mrr):
(JSC::X86Assembler::vpinsrd_i8rrr):
(JSC::X86Assembler::vpinsrq_i8mrr):
(JSC::X86Assembler::vpinsrq_i8rrr):
(JSC::X86Assembler::vinsertps_i8rrr):
(JSC::X86Assembler::vmovddup_rr):
(JSC::X86Assembler::vmovddup_mr):
(JSC::X86Assembler::vmovapd_rr):
(JSC::X86Assembler::vbroadcastss_mr):
(JSC::X86Assembler::vmovq_rr):
(JSC::X86Assembler::vpunpcklqdq_rrr):
(JSC::X86Assembler::vunpcklpd_rrr):
(JSC::X86Assembler::vpextrb_i8rr):
(JSC::X86Assembler::vpextrb_i8rm):
(JSC::X86Assembler::vpextrw_i8rr):
(JSC::X86Assembler::vpextrw_i8rm):
(JSC::X86Assembler::vpextrd_i8rr):
(JSC::X86Assembler::vpextrd_i8rm):
(JSC::X86Assembler::vpextrq_i8rr):
(JSC::X86Assembler::vpextrq_i8rm):
(JSC::X86Assembler::vshufps_i8rrr):
(JSC::X86Assembler::vshufpd_i8rrr):
(JSC::X86Assembler::vpshuflw_i8rr):
(JSC::X86Assembler::vpshufd_i8rr):
(JSC::X86Assembler::vpackusdw_rrr):
(JSC::X86Assembler::vpmovsxbw_rr):
(JSC::X86Assembler::vpmovzxbw_rr):
(JSC::X86Assembler::vpmovsxwd_rr):
(JSC::X86Assembler::vpmovzxwd_rr):
(JSC::X86Assembler::vpmovsxdq_rr):
(JSC::X86Assembler::vpmovzxdq_rr):
(JSC::X86Assembler::vmovaps_rr):
(JSC::X86Assembler::vmovshdup_rr):
(JSC::X86Assembler::vmovhlps_rrr):
(JSC::X86Assembler::vmovsd_rrr):
(JSC::X86Assembler::vpblendw_i8rrr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
(JSC::X86Assembler::pinsrb_rr): Deleted.
(JSC::X86Assembler::pinsrw_rr): Deleted.
(JSC::X86Assembler::pinsrd_rr): Deleted.
(JSC::X86Assembler::pinsrq_rr): Deleted.
(JSC::X86Assembler::insertps_rr): Deleted.
(JSC::X86Assembler::pextrb_rr): Deleted.
(JSC::X86Assembler::pextrw_rr): Deleted.
(JSC::X86Assembler::pextrd_rr): Deleted.
(JSC::X86Assembler::pextrq_rr): Deleted.
(JSC::X86Assembler::pshufd_rr): Deleted.
(JSC::X86Assembler::pshuflw_rr): Deleted.
(JSC::X86Assembler::shufps_rr): Deleted.
(JSC::X86Assembler::shufpd_rr): Deleted.
(JSC::X86Assembler::vpextrb_rr): Deleted.
(JSC::X86Assembler::vpextrw_rr): Deleted.
(JSC::X86Assembler::vpextrd_rr): Deleted.
(JSC::X86Assembler::vpextrq_rr): Deleted.
(JSC::X86Assembler::vshufps_rrr): Deleted.
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDSwizzleHelperX86):
(JSC::Wasm::AirIRGenerator::addSIMDV_VV):
(JSC::Wasm::AirIRGenerator::addSIMDShuffle):
(JSC::Wasm::AirIRGenerator::addSIMDLoadSplat):
(JSC::Wasm::AirIRGenerator::addSIMDSwizzle): Deleted.

Canonical link: https://commits.webkit.org/257818@main
